make osd device configurable + adding support for tinyOSD

fishpepper opened the pull request:
Do not merge this, for now this is only for discussion.

The idea is to allow various (runtime selectable) osd devices in the future.
For now osd_device can be  none, max7456, msp, tinyOSD

Additionally this adds support for my serial tinyOSD protocol. See issue #3397 for more info.

Example for enabling tinyOSD on serial3 via CLI:
```
set osd_device = tinyOSD
serial 2 32768 115200 115200 115200 115200
feature osd
save
```
This will output the osd datastream on serial port 3 @115200 baud.

DieHertz commented:
Seems like you've merged instead of rebasing

azolyoung commented:
It's a good idea! If can setup OSD device with GUI instead of CLI, it'll be perfect!

sirPerna commented:
Did you guys consider to use CRSF for transportation over serial as the driver is already inside the code?

digitalentity commented:
@sirPerna you mean add extra CRSF messages? Might be good thing to consider - minimal overhead (3 bytes per frame), same checksumming. @fishpepper openTCO is essentially CRSF - just add your new commands to CRSF telemetry.

sirPerna commented:
@digitalentity Yes exactly. It's really easy to add extra messages.

fishpepper commented:
thanks for the crsf pointer! I don't think integrating the camera/vtx/osd stuff into crsf would be the optimal way. It is meant for different purposes and e.g. the timer based sending of single telemetry packets is to different to what I have in mind.
But the way the crsf protocol was implemented looks really nice. Thanks! I restructured my stuff to use the (very nice) streambuffer implementation that crsf uses. Looks much more readable now :)

azolyoung commented:
@fishpepper 

Hi, I'm integrating openCTO into my split osd feature, and I've rebase you changes from the osd_device branch in your cleanflight repository.
And in opentco.c, the protocol is implement with the global serial port, buffer and lock flag, that's means: the openTCO only work with one device, 
I can't found how to make a pull request to you betaflight repository, so I make  little changes in opentco.c, opentco.h and tinyosd.c, you can found it in my cf repository:https://github.com/azolyoung/cleanflight/commit/f58124e83db9a7120ca01d1e74cc1fce7d9c1ae5

Hope to work with you to improve this change (so that the implementation of openTCO support multiple devices)

fishpepper commented:
I updated the protocol a bit and added register read/write commands.

openTCO devices can now share a single serial port (e.g. having vtx and osd on the same port).
Single port per device is also possible, the code will scan all serialports that have FUNCTION_OPENTCO set for valid devices.

OSD devices are now checked for their capabilities by a setting the enable flags of all features and reading back what was activated.

DieHertz commented:
Awesome, I'll have to use your PR to test OpenVTX

fishpepper commented:
I am going to add a suggestion for the openTCO protocol and implementation for vtx devices soon :)
Make sure to use the patched configurator from my github fork as well. I added user changable brightness/invert settings for max7456 based OSDs as well.
In order to activate openTCO: go to ports, select openTCO at peripherals and save (115200baud), in configuration activate OSD and the go to the OSD tab and select openTCO as osd device (this should probably be moved to configuration).
Or via CLI:
```
feature osd
serial x 32768 115200 115200 115200 115200
set osd_device=opentco
save
```

DieHertz commented:
OpenVTX has no osd capability at this point so I can't test anything osd related, but I'll do my best with VTX code :-)

fishpepper commented:
thanks, added 600mW
The current implementation also reports back which levels are supported out of the possibilities listed.
I am not sure (yet) how to implement it in a more general way...

DieHertz commented on src/main/drivers/opentco.c:
```diff
@@ -0,0 +1,406 @@
+
+    while (portConfig != NULL) {
+        // extract baudrate
+        uint32_t baudrate = baudRates[portConfig->blackbox_baudrateIndex];
```
Why use `blackbox_baudrateIndex`?

mikeller commented:
@fishpepper: What is the status of this?
